python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll

```diff
@@ -9,3 +9,136 @@ private import semmle.python.dataflow.new.TaintTracking
 private import semmle.python.dataflow.new.RemoteFlowSources
 private import experimental.semmle.python.Concepts
 private import semmle.python.ApiGraphs
+
+private module NoSQL {
+  /** Gets a reference to a `MongoClient` DB. */
+  private API::Node mongoClientInstance() {
+    result = API::moduleImport("pymongo").getMember("MongoClient").getReturn() or
+    result =
+      API::moduleImport("flask_mongoengine")
+          .getMember("MongoEngine")
+          .getReturn()
+          .getMember("get_db")
+          .getReturn() or
+    result = API::moduleImport("mongoengine").getMember("connect").getReturn() or
+    result = API::moduleImport("flask_pymongo").getMember("PyMongo").getReturn()
+  }
+
+  /** Gets a reference to a `MongoClient` DB. */
+  private DataFlow::LocalSourceNode mongoClientDB(DataFlow::TypeTracker t) {
+    t.start() and
+    (
+      exists(SubscriptNode subscript | result.asCfgNode() = subscript |
+        subscript.getObject() = mongoClientInstance().getAUse().asCfgNode()
+      )
+      or
+      result.(DataFlow::AttrRead).getObject() = mongoClientInstance().getAUse()
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = mongoClientDB(t2).track(t2, t))
+  }
+
+  /** Gets a reference to a `MongoClient` DB. */
+  private DataFlow::Node mongoClientDB() {
+    mongoClientDB(DataFlow::TypeTracker::end()).flowsTo(result)
+  }
+
+  /** Gets a reference to a `MongoClient` collection. */
+  private DataFlow::LocalSourceNode mongoClientCollection(DataFlow::TypeTracker t) {
+    t.start() and
+    (
+      exists(SubscriptNode subscript | result.asCfgNode() = subscript |
+        subscript.getObject() = mongoClientDB().asCfgNode()
+      )
+      or
+      result.(DataFlow::AttrRead).getObject() = mongoClientDB()
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = mongoClientCollection(t2).track(t2, t))
+  }
+
+  /** Gets a reference to a `MongoClient` collection. */
+  private DataFlow::Node mongoClientCollection() {
+    mongoClientCollection(DataFlow::TypeTracker::end()).flowsTo(result)
+  }
+
+  /** This class represents names of find_* relevant MongoClient collection level operation methods. */
+  private class MongoClientMethodNames extends string {
+    MongoClientMethodNames() {
+      // the find_one_or_404 method is only found in the Pymongo Flask library.  
+      this in [
+          "find", "find_raw_batches", "find_one", "find_one_and_delete", "find_and_modify",
+          "find_one_and_replace", "find_one_and_update", "find_one_or_404"
+        ]
+    }
+  }
+
+  /** Gets a reference to a `MongoClient` Collection method. */
+  DataFlow::Node mongoClientMethod() {
+    result.(DataFlow::AttrRead).getAttributeName() instanceof MongoClientMethodNames and
+    (
+      result.(DataFlow::AttrRead).getObject() = mongoClientCollection() or
+      result.(DataFlow::AttrRead) = mongoClientCollection()
+    )
+  }
```

Comment:
I don't understand this. Would the following ever be allowed? :flushed: 

```py
client = pymongo.MongoClient(...)
client.mydb.find(...)  # <-- I'm surprised if this one is allowed :O
client.mydb.mycollection.find(...)
```

at least looking at the [docs](https://pymongo.readthedocs.io/en/stable/api/pymongo/database.html#pymongo.database.Database), that does notseem to be the case.. I would think the line `result.(DataFlow::AttrRead) = mongoClientCollection()` should be removed then (?)

Reply:
This is one of the frustrating things with both MongoEngine and Flask MongoEngine. There are multiple ways to retrieve a database. In this case,  a developer can do the following:

```python
import mongoengine as me
db = me.get_db('name_of_db')
data = db.name_of_collection.find({'name': malicious_input})
```

This predicate checks for those scenarios. I reviewed the tests I just wrote and I have three tests covering this scenario (albeit the tests are checking other things, just not this)

- mongoengine_get_db_via_connection_bad.py
- mongoengine_get_db_bad.py
- flask_mongoengine_get_db_bad.py